Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,700 @@
+      {
+        ref->tcp_socket.send(data, size);
+        ref->ack_num += size;
+      }
+
+      // send ack
+      BuildAckFrame(ref, m_in_frame.get());
+
+      WriteToQueue(m_in_frame.get(), 0x36);
+    }
+    // update windows size
+    ref->window_size = ntohs(tcpdata->window_size);
+
+    // clear any ack data
+    if (tcpdata->properties & TCP_FLAG_ACK)
+    {
+      const u32 ack_num = ntohl(tcpdata->acknowledgement_number);
+      for (auto& tcp_buf : ref->tcp_buffers)
+      {
+        if (!tcp_buf.used || tcp_buf.seq_id >= ack_num)
+          continue;
+        Common::TCPHeader* tcppart = (Common::TCPHeader*)&tcp_buf.data[0x22];
+        const u32 seq_end =
+            tcp_buf.seq_id + tcp_buf.data_size - ((tcppart->properties & 0xf0) >> 2) - 34;
+        if (seq_end <= ack_num)
+        {
+          tcp_buf.used = false;  // confirmed data received
+          if (!ref->ready && !ref->tcp_buffers[0].used)
+            ref->ready = true;
+          continue;
+        }
+        // partial data, adjust the packet for next ack
+        const u16 ack_size = ack_num - tcp_buf.seq_id;
+        const u16 new_data_size = tcp_buf.data_size - 0x36 - ack_size;
+        std::memmove(&tcp_buf.data[0x36], &tcp_buf.data[0x36 + ack_size], new_data_size);
+        tcp_buf.data_size -= ack_size;
+        tcp_buf.seq_id += ack_size;
+        tcppart->sequence_number = htonl(tcp_buf.seq_id);
+        Common::IPv4Header* ippart = (Common::IPv4Header*)&tcp_buf.data[14];
+        ippart->total_len = htons(tcp_buf.data_size - 14);
+        tcppart->checksum = 0;
+        tcppart->checksum = Common::ComputeTCPNetworkChecksum(ref->from, ref->to, tcppart,
+                                                              new_data_size + 20, IPPROTO_TCP);
+      }
+    }
+  }
+}
+
+/// <summary>
+/// This is a litle hack, Mario Kart open some UDP port
+/// and listen to it. We open it on our side manualy.
+/// </summary>
+void CEXIETHERNET::BuiltInBBAInterface::InitUDPPort(u16 port)
+{
+  StackRef* ref = GetAvaibleSlot(htons(port));
+  if (ref == nullptr || ref->ip != 0)
+    return;
+  ref->ip = 0x08080808;  // change for ip
+  ref->local = htons(port);
+  ref->remote = htons(port);
+  ref->type = 17;
+  ref->bba_mac = *(Common::MACAddress*)&m_eth_ref->mBbaMem[BBA_NAFR_PAR0];
+  ref->my_mac = fake_mac;
+  ref->from.sin_addr.s_addr = 0;
+  ref->from.sin_port = htons(port);
+  ref->to.sin_addr.s_addr = m_current_ip;
+  ref->to.sin_port = htons(port);
+  ref->udp_socket.setBlocking(false);
+  if (ref->udp_socket.bind(port) != sf::Socket::Done)
+  {
+    ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
+    return;
+  }
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::HandleUDPFrame(Common::EthernetHeader* hwdata,
+                                                       Common::IPv4Header* ipdata,
+                                                       Common::UDPHeader* udpdata, u8* data)
+{
+  sf::IpAddress target;
+
+  if (*(u32*)ipdata->destination_addr == 0)
```

Comment:
You're doing this cast to pointer -> deref pointer all over the place, that is UB. Please use Common::BitCast instead.

Reply:
I understand that using pointer is prohibited, this is prety much the first real c++ code i make so i need your help on this. I need a way to access various layer of data from the BBA buffer given to me by the NetworkInterface. 

So im getting a buffer of bytes, it start with 2 or more headers with variable sizes. i need a way to map them so i can use theire information to do proper NAT forwarding. Something similar is also needed to rebuild the frame data that will be pushed to the BBA Buffer.

Reply:
So fully explaining the C++ Standard and type punning and strict aliasing would take hours, but in short: Even though it often works, a cast like this violates C++ rules and may cease to work with a different compiler or different compiler optimizations.

If you replace this with something like a BitCast, so, eg.

~~~
  if (*(u32*)ipdata->destination_addr == 0)
    *(u32*)ipdata->destination_addr = m_router_ip;
~~~

becomes

~~~
  if (Common::BitCast<u32>(ipdata->destination_addr) == 0)
    ipdata->destination_addr = Common::BitCast<std::array<u8, IPV4_ADDR_LEN>>(m_router_ip);
~~~

it does the same thing, but is guaranteed to keep working regardless of compiler, settings, etc. -- though note that you may need to wait for #10691 to be merged for the second line of that to work.